### PR TITLE
[SPARK-8138] [SQL] Improves error message when conflicting partition columns are found

### PR DESCRIPTION
**Pull request by liancheng**
This PR improves the error message shown when conflicting partition column names are detected.  This can be particularly annoying and confusing when there are a large number of partitions while a handful of them happened to contain unexpected temporary file(s).  Now all suspicious directories are listed as below:

```
java.lang.AssertionError: assertion failed: Conflicting partition column names detected:

        Partition column name list #0: b, c, d
        Partition column name list #1: b, c
        Partition column name list #2: b

For partitioned table directories, data files should only live in leaf directories. Please check the following directories for unexpected files:

        file:/tmp/foo/b=0
        file:/tmp/foo/b=1
        file:/tmp/foo/b=1/c=1
        file:/tmp/foo/b=0/c=0
```

### CODE REVIEW

**Comment by liancheng** [comment]
cc @rxin

**Comment by yhuai** [comment]
Let's create a jira for this.

**Comment by liancheng** [comment]
Filed SPARK-8138 for this and updated PR title.

**Comment by liancheng** [comment]
@yhuai As discussed offline, now we give a more descriptive and help message with a list of all suspicious non-leaf partition directories.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/PartitioningUtils.scala [hunk 1]
```diff
@@ -179,19 +179,41 @@ private[sql] object PartitioningUtils {
...
+    if (pathsWithPartitionValues.isEmpty) {
       Seq.empty
     } else {
-      assert(distinctPartitionsColNames.size == 1, {
-        val list = distinctPartitionsColNames.mkString("\t", "\n\t", "")
-        s"Conflicting partition column names detected:\n$list"
-      })
+      val distinctPartColNames = pathsWithPartitionValues.map(_._2.columnNames).distinct
+
+      def listConflictingPartitionColumns: String = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/PartitioningUtils.scala]: move this out, make it accept arguments in the form of collections, and write unit test for this function.

this function is way too complicated to not have unit tests.

**Comment by liancheng** [comment]
retest this please

**Comment by liancheng** [comment]
Merging to master.

**Pull request closed without merging**